// vite-plugin-html-minify/index.js
import { minify } from 'html-minifier-terser'
import { createFilter } from '@rollup/pluginutils'

const htmlFilter = createFilter(['**/*.html'])

const vitePluginHtmlMinify = (options = {}) => {
    return {
        name: 'html-minify',
        enforce: 'post',
        async generateBundle(_, outBundle) {
            for (const bundle of Object.values(outBundle)) {
                if (
                    bundle.type === 'asset' &&
                    htmlFilter(bundle.fileName) &&
                    typeof bundle.source === 'string'
                ) {
                    bundle.source = await minify(
                        bundle.source,
                        Object.assign(
                            {},
                            {
                                collapseWhitespace: true,
                                keepClosingSlash: true,
                                removeComments: true,
                                removeRedundantAttributes: true,
                                removeScriptTypeAttributes: true,
                                removeStyleLinkTypeAttributes: true,
                                useShortDoctype: true,
                                minifyCSS: true
                            },
                            options
                        )
                    )
                }
            }
        }
    }
}
export {vitePluginHtmlMinify}
export default vitePluginHtmlMinify
